src/net/ossl.cc

```diff
@@ -173,4 +188,262 @@ tls::dh_params::~dh_params() {
+        // the store represents the root and intermediate chain. Exposed for later
+        // use when the ssl socket is created
+        X509* x509_tmp = nullptr;
+        if(nullptr == parse_x509_cert(cert, fmt, &x509_tmp)){
+            throw ossl_error();
+        }
+        auto x509_cert = x509_ptr(x509_tmp);
+        bio_ptr key_bio(BIO_new_mem_buf(key.begin(), key.size()));
+        auto pkey_temp = EVP_PKEY_new();
+        if (nullptr == PEM_read_bio_PrivateKey(key_bio.get(), &pkey_temp, nullptr, nullptr)) {
+            EVP_PKEY_free(pkey_temp);
+            throw ossl_error();
+        }
+        auto pkey = evp_pkey_ptr(pkey_temp);
+        _ck_pair = std::make_optional(cert_key_pair{.cert = std::move(x509_cert), .key = std::move(pkey)});
+    }
+
+    void set_simple_pkcs12(const blob& b, x509_crt_format fmt, const sstring& password) {
+        // Load the PKCS12 file
+        bio_ptr bio(BIO_new_mem_buf(b.begin(), b.size()));
+        PKCS12 *p12_tmp = nullptr;
+        if(nullptr == d2i_PKCS12_bio(bio.get(), &p12_tmp)) {
+            throw ossl_error();
+        }
+        auto p12 = pkcs12(p12_tmp);
+        // Extract the certificate and private key from PKCS12, using provided password
+        EVP_PKEY *pkey = nullptr;
+        X509 *cert = nullptr;
+        if (!PKCS12_parse(p12.get(), password.c_str(), &pkey, &cert, nullptr)) {
+            throw ossl_error();
+        }
+        X509_STORE_add_cert(*this, cert);
+        EVP_PKEY_free(pkey);
+    }
+
+    void dh_params(const tls::dh_params& dh) {
+        auto cpy = std::make_shared<tls::dh_params::impl>(*dh._impl);
+        _dh_params = std::move(cpy);
+    }
+
+    void set_client_auth(client_auth ca) {
+        _client_auth = ca;
+    }
+    client_auth get_client_auth() const {
+        return _client_auth;
+    }
+
+    void set_priority_string(const sstring& prio) {
+        // if (SSL_set_cipher_list(_ssl.get(), prio) != 1) {
+        // if (SSL_CTX_set_cipher_list(_ctx.get(), prio) != 1) {
+        //     // Handle error
+        // }
+    }
+
+    void set_dn_verification_callback(dn_callback cb) {
+        _dn_callback = std::move(cb);
+    }
+
+
+    operator X509_STORE*() const { return _creds.get(); }
+
+    const std::optional<cert_key_pair>& get_cert_key_pair() const {
+        return _ck_pair;
+    }
+
+    future<> set_system_trust() {
+        /// TODO: Must have access to the ssl context to call this method
+        /// SSL_CTX_set0_verify_cert_store(_ssl.get(), _creds.get());
+        // _load_system_trust = false;
+        return make_ready_future<>();
+    }
+
+    // bool need_load_system_trust() const {
+    //     return _load_system_trust;
+    // }
+
+private:
+    friend class credentials_builder;
+
+    x509_store_ptr _creds;
+
+    // Must retain the certs and CRLs as X509_STORE_add_cert will not take
+    // ownership of the pointers when certs are added to the store
```

Comment:
The wording in their documentation is weird and it tripped me up a little bit, but after examining the source, you don't have to do this.

When the object is added to the store, the internal reference count is incremented, meaning you're free to call "free" on the object you have but it won't be destroyed.  "free" just decrements the reference count until it's 0 and then it's cleaned up.

The store will call free on the objects it holds when it is free'ed.